pkg/gadgets/tcptracer/tracer/core/bpf/tcptracer.bpf.c

```diff
@@ -0,0 +1,328 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2022 Microsoft Corporation
+//
+// Based on tcptracer(8) from BCC by Kinvolk GmbH and
+// tcpconnect(8) by Anton Protopopov
+#include <vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+#include "tcptracer.h"
+
+const volatile uid_t filter_uid = -1;
+const volatile pid_t filter_pid = 0;
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+/* tcp_set_state doesn't run in the context of the process that initiated the
+ * connection so we need to store a map TUPLE -> PID to send the right PID on
+ * the event
+ */
+struct tuple_key_t {
+	union {
+		__u32 saddr_v4;
+		unsigned __int128 saddr_v6;
+	};
+	union {
+		__u32 daddr_v4;
+		unsigned __int128 daddr_v6;
+	};
+	u16 sport;
+	u16 dport;
+	u32 netns;
+};
+
+struct pid_comm_t {
+	u64 pid;
+	char comm[TASK_COMM_LEN];
+	u32 uid;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct tuple_t);
+	__type(value, struct pid_comm_t);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} tuplepid SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct sock *);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} sockets SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+
+static __always_inline bool
+fill_tuple(struct tuple_key_t *tuple, struct sock *sk, int family)
+{
+	struct inet_sock *sockp = (struct inet_sock *)sk;
+
+	if (family == AF_INET) {
+		BPF_CORE_READ_INTO(&tuple->saddr_v4, sk, __sk_common.skc_rcv_saddr);
+		if (tuple->saddr_v4 == 0)
+			return false;
+
+		BPF_CORE_READ_INTO(&tuple->daddr_v4, sk, __sk_common.skc_daddr);
+		if (tuple->daddr_v4 == 0)
+			return false;
+	} else if (family == AF_INET6) {
+		BPF_CORE_READ_INTO(&tuple->saddr_v6, sk,
+				   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+		if (tuple->saddr_v6 == 0)
+			return false;
+		BPF_CORE_READ_INTO(&tuple->daddr_v6, sk,
+				   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+		if (tuple->daddr_v6 == 0)
+			return false;
+	}
```

Comment:
Can the case where `family` is neither `AF_INET` nor `AFTER_INET6` happen?
If so, what about adding an `else` to deal with it?

Reply:
I changed to use a switch statement and I'm handling the default case too (just in case).